cmd/holo-files/internal/impl/resource_patch.go

```diff
@@ -0,0 +1,111 @@
+/*******************************************************************************
+*
+* Copyright 2017-2018 Luke Shumaker <lukeshu@parabola.nu>
+*
+* This file is part of Holo.
+*
+* Holo is free software: you can redistribute it and/or modify it under the
+* terms of the GNU General Public License as published by the Free Software
+* Foundation, either version 3 of the License, or (at your option) any later
+* version.
+*
+* Holo is distributed in the hope that it will be useful, but WITHOUT ANY
+* WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+* A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License along with
+* Holo. If not, see <http://www.gnu.org/licenses/>.
+*
+*******************************************************************************/
+
+package impl
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/holocm/holo/cmd/holo-files/internal/common"
+)
+
+// Patchfile is a Resource that is a `patch(1)` file that edits the
+// current version of the entity.
+type Patchfile struct{ rawResource }
+
+// ApplicationStrategy implements the Resource interface.
+func (resource Patchfile) ApplicationStrategy() string { return "patch" }
+
+// DiscardsPreviousBuffer implements the Resource interface.
+func (resource Patchfile) DiscardsPreviousBuffer() bool { return false }
+
+// ApplyTo implements the Resource interface.
+func (resource Patchfile) ApplyTo(entityBuffer common.FileBuffer) (common.FileBuffer, error) {
+	// `patch` requires that the file it's operating on be a real
+	// file (not a pipe).  So, we'll write entityBuffer to a
+	// temporary file, run `patch`, then read it back.
+
+	// We really only normally need 1 temporary file, but:
+	//  1. since common.FileBuffer.Write removes the file and then
+	//     re-creates it, that's a bit racy
+	//  2. The only way to limit patch to operating on a single
+	//     file is to name that file on the command line, but
+	//     doing that prevents it from unlinking the file, which
+	//     prevents type changes.
+	//
+	// Using a temporary directory lets us easily work around both
+	// of these issues.  Unfortunately, this allows the patch to
+	// create new files other than the one for the entity we are
+	// applying.  However, it can't escape the temporary
+	// directory, so we'll just "allow" that, and document that we
+	// ignore those files.
+	targetDir, err := ioutil.TempDir(os.Getenv("HOLO_CACHE_DIR"), "patch-target.")
+	if err != nil {
+		return common.FileBuffer{}, err
+	}
+	defer os.RemoveAll(targetDir)
+	targetPath := filepath.Join(targetDir, filepath.Base(entityBuffer.Path))
```

Comment:
`targetPath` is a bit confusing since we already have a "target path" in our jargon at some other point. Maybe `workDir` and `workPath`, or `tempDir` and `tempPath`?